File: java/IceStorm/clock/Subscriber.java
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import com.zeroc.demos.IceStorm.clock.Demo.*;

public class Subscriber
{
    public static class ClockI implements Clock
    {
        @Override
        public void tick(String date, com.zeroc.Ice.Current current)
        {
            System.out.println(date);
        }
    }

    public static void main(String[] args)
    {
        int status = 0;
        java.util.List<String> extraArgs = new java.util.ArrayList<String>();

        com.zeroc.Ice.Communicator communicator = com.zeroc.Ice.Util.initialize(args, "config.sub", extraArgs);
        communicator.getProperties().setProperty("Ice.Default.Package", "com.zeroc.demos.IceStorm.clock");
        //
        // Destroy communicator during JVM shutdown
        //
        Thread destroyHook = new Thread(() -> communicator.destroy());
        Runtime.getRuntime().addShutdownHook(destroyHook);

        try
        {
            status = run(communicator, destroyHook, extraArgs.toArray(new String[extraArgs.size()]));
        }
        catch(Exception ex)
        {
            ex.printStackTrace();
            status = 1;
        }

        if(status != 0)
        {
            System.exit(status);
        }
        //
        // Else the application waits for Ctrl-C to destroy the communicator
        //
    }

    public static void usage()
    {
        System.out.println("Usage: [--batch] [--datagram|--twoway|--ordered|--oneway] " +
                           "[--retryCount count] [--id id] [topic]");
    }

    private static int run(com.zeroc.Ice.Communicator communicator, Thread destroyHook, String[] args)
    {
        args = communicator.getProperties().parseCommandLineOptions("Clock", args);

        String topicName = "time";
        String option = "None";
        boolean batch = false;
        String id = null;
        String retryCount = null;
        int i;
        for(i = 0; i < args.length; ++i)
        {
            String oldoption = option;
            if(args[i].equals("--datagram"))
            {
                option = "Datagram";
            }
            else if(args[i].equals("--twoway"))
            {
                option = "Twoway";
            }
            else if(args[i].equals("--ordered"))
            {
                option = "Ordered";
            }
            else if(args[i].equals("--oneway"))
            {
                option = "Oneway";
            }
            else if(args[i].equals("--batch"))
            {
                batch = true;
            }
            else if(args[i].equals("--id"))
            {
                ++i;
                if(i >= args.length)
                {
                    usage();
                    return 1;
                }
                id = args[i];
            }
            else if(args[i].equals("--retryCount"))
            {
                ++i;
                if(i >= args.length)
                {
                    usage();
                    return 1;
                }
                retryCount = args[i];
            }
            else if(args[i].startsWith("--"))
            {
                usage();
                return 1;
            }
            else
            {
                topicName = args[i++];
                break;
            }

            if(!oldoption.equals(option) && !oldoption.equals("None"))
            {
                usage();
                return 1;
            }
        }

        if(i != args.length)
        {
            usage();
            return 1;
        }

        if(batch && (option.equals("Twoway") || option.equals("Ordered")))
        {
            System.err.println("batch can only be set with oneway or datagram");
            return 1;
        }

        com.zeroc.IceStorm.TopicManagerPrx manager = com.zeroc.IceStorm.TopicManagerPrx.checkedCast(
            communicator.propertyToProxy("TopicManager.Proxy"));
        if(manager == null)
        {
            System.err.println("invalid proxy");
            return 1;
        }

        //
        // Retrieve the topic.
        //
        com.zeroc.IceStorm.TopicPrx topic;
        try
        {
            topic = manager.retrieve(topicName);
        }
        catch(com.zeroc.IceStorm.NoSuchTopic e)
        {
            try
            {
                topic = manager.create(topicName);
            }
            catch(com.zeroc.IceStorm.TopicExists ex)
            {
                System.err.println("temporary failure, try again.");
                return 1;
            }
        }

        com.zeroc.Ice.ObjectAdapter adapter = communicator.createObjectAdapter("Clock.Subscriber");

        //
        // Add a servant for the Ice object. If --id is used the
        // identity comes from the command line, otherwise a UUID is
        // used.
        //
        // id is not directly altered since it is used below to detect
        // whether subscribeAndGetPublisher can raise
        // AlreadySubscribed.
        //
        com.zeroc.Ice.Identity subId = new com.zeroc.Ice.Identity(id, "");
        if(subId.name == null)
        {
            subId.name = java.util.UUID.randomUUID().toString();
        }
        com.zeroc.Ice.ObjectPrx subscriber = adapter.add(new ClockI(), subId);

        //
        // Activate the object adapter before subscribing.
        //
        adapter.activate();

        java.util.Map<String, String> qos = new java.util.HashMap<>();
        if(retryCount != null)
        {
            qos.put("retryCount", retryCount);
        }
        //
        // Set up the proxy.
        //
        if(option.equals("Datagram"))
        {
            if(batch)
            {
                subscriber = subscriber.ice_batchDatagram();
            }
            else
            {
                subscriber = subscriber.ice_datagram();
            }
        }
        else if(option.equals("Twoway"))
        {
            // Do nothing to the subscriber proxy. Its already twoway.
        }
        else if(option.equals("Ordered"))
        {
            // Do nothing to the subscriber proxy. Its already twoway.
            qos.put("reliability", "ordered");
        }
        else if(option.equals("Oneway") || option.equals("None"))
        {
            if(batch)
            {
                subscriber = subscriber.ice_batchOneway();
            }
            else
            {
                subscriber = subscriber.ice_oneway();
            }
        }

        try
        {
            topic.subscribeAndGetPublisher(qos, subscriber);
        }
        catch(com.zeroc.IceStorm.AlreadySubscribed e)
        {
            // This should never occur when subscribing with an UUID
            if(id == null)
            {
                e.printStackTrace();
                return 1;
            }
            System.out.println("reactivating persistent subscriber");
        }
        catch(com.zeroc.IceStorm.InvalidSubscriber e)
        {
            e.printStackTrace();
            return 1;
        }
        catch(com.zeroc.IceStorm.BadQoS e)
        {
            e.printStackTrace();
            return 1;
        }

        //
        // Replace the shutdown hook to unsubscribe during JVM shutdown
        //
        final com.zeroc.IceStorm.TopicPrx topicF = topic;
        final com.zeroc.Ice.ObjectPrx subscriberF = subscriber;
        Runtime.getRuntime().addShutdownHook(new Thread(() ->
        {
            try
            {
                topicF.unsubscribe(subscriberF);
            }
            finally
            {
                communicator.destroy();
            }
        }));
        Runtime.getRuntime().removeShutdownHook(destroyHook); // remove old destroy-only shutdown hook

        return 0;
    }
}

File: csharp/IceStorm/clock/Subscriber.cs
// Copyright (c) ZeroC, Inc.

using Demo;
using System.Diagnostics;

internal class Subscriber
{
    internal class ClockI : ClockDisp_
    {
        public override void tick(string date, Ice.Current current)
        {
            Console.WriteLine(date);
        }
    }

    public static int Main(string[] args)
    {
        int status = 0;

        try
        {
            // using statement - communicator is automatically destroyed
            // at the end of this statement
            using var communicator = Ice.Util.initialize(ref args, "config.sub");
            // Shutdown the communicator on Ctrl+C or Ctrl+Break
            // (always use Cancel = true with shutdown)
            Console.CancelKeyPress += (sender, eventArgs) =>
            {
                eventArgs.Cancel = true;
                communicator.shutdown();
            };

            status = Run(communicator, args);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(ex);
            status = 1;
        }

        return status;
    }

    private static int Run(Ice.Communicator communicator, string[] args)
    {
        args = communicator.getProperties().parseCommandLineOptions("Clock", args);

        string topicName = "time";
        string option = "None";
        bool batch = false;
        string id = null;
        string retryCount = null;
        int i;
        for (i = 0; i < args.Length; ++i)
        {
            string oldoption = option;
            if (args[i].Equals("--datagram"))
            {
                option = "Datagram";
            }
            else if (args[i].Equals("--twoway"))
            {
                option = "Twoway";
            }
            else if (args[i].Equals("--ordered"))
            {
                option = "Ordered";
            }
            else if (args[i].Equals("--oneway"))
            {
                option = "Oneway";
            }
            else if (args[i].Equals("--batch"))
            {
                batch = true;
            }
            else if (args[i].Equals("--id"))
            {
                ++i;
                if (i >= args.Length)
                {
                    Usage();
                    return 1;
                }
                id = args[i];
            }
            else if (args[i].Equals("--retryCount"))
            {
                ++i;
                if (i >= args.Length)
                {
                    Usage();
                    return 1;
                }
                retryCount = args[i];
            }
            else if (args[i].StartsWith("--"))
            {
                Usage();
                return 1;
            }
            else
            {
                topicName = args[i++];
                break;
            }

            if (!oldoption.Equals(option) && !oldoption.Equals("None"))
            {
                Usage();
                return 1;
            }
        }

        if (i != args.Length)
        {
            Usage();
            return 1;
        }

        if (batch && (option.Equals("Twoway") || option.Equals("Ordered")))
        {
            Console.WriteLine("batch can only be set with oneway or datagram");
            return 1;
        }

        IceStorm.TopicManagerPrx manager = IceStorm.TopicManagerPrxHelper.checkedCast(
            communicator.propertyToProxy("TopicManager.Proxy"));
        if (manager == null)
        {
            Console.WriteLine("invalid proxy");
            return 1;
        }

        // Retrieve the topic.
        IceStorm.TopicPrx topic;
        try
        {
            topic = manager.retrieve(topicName);
        }
        catch (IceStorm.NoSuchTopic)
        {
            try
            {
                topic = manager.create(topicName);
            }
            catch (IceStorm.TopicExists)
            {
                Console.WriteLine("temporary error. try again.");
                return 1;
            }
        }

        Ice.ObjectAdapter adapter = communicator.createObjectAdapter("Clock.Subscriber");
        // Add a servant for the Ice object. If --id is used the
        // identity comes from the command line, otherwise a UUID is
        // used.
        //
        // id is not directly altered since it is used below to
        // detect whether subscribeAndGetPublisher can raise
        // AlreadySubscribed.
        Ice.Identity subId = new Ice.Identity(id, "");
        if (subId.name == null)
        {
            subId.name = Guid.NewGuid().ToString();
        }
        Ice.ObjectPrx subscriber = adapter.add(new ClockI(), subId);

        // Activate the object adapter before subscribing.
        adapter.activate();

        Dictionary<string, string> qos = new Dictionary<string, string>();
        if (retryCount != null)
        {
            qos["retryCount"] = retryCount;
        }
        // Set up the proxy.
        if (option.Equals("Datagram"))
        {
            if (batch)
            {
                subscriber = subscriber.ice_batchDatagram();
            }
            else
            {
                subscriber = subscriber.ice_datagram();
            }
        }
        else if (option.Equals("Twoway"))
        {
            // Do nothing to the subscriber proxy. Its already twoway.
        }
        else if (option.Equals("Ordered"))
        {
            // Do nothing to the subscriber proxy. Its already twoway.
            qos["reliability"] = "ordered";
        }
        else if (option.Equals("Oneway") || option.Equals("None"))
        {
            if (batch)
            {
                subscriber = subscriber.ice_batchOneway();
            }
            else
            {
                subscriber = subscriber.ice_oneway();
            }
        }

        try
        {
            topic.subscribeAndGetPublisher(qos, subscriber);
        }
        catch (IceStorm.AlreadySubscribed)
        {
            // This should never occur when subscribing with an UUID
            Debug.Assert(id != null);
            Console.WriteLine("reactivating persistent subscriber");
        }

        communicator.waitForShutdown();
        topic.unsubscribe(subscriber);

        return 0;
    }

    private static void Usage()
    {
        Console.WriteLine("Usage: [--batch] [--datagram|--twoway|--ordered|--oneway] " +
                          "[--retryCount count] [--id id] [topic]");
    }
}
